### PR TITLE
[CDAP-17549] Adds pipeline level comments + handles pipelines in deployed view

### PR DESCRIPTION
**Pull request by ajainarayanan**
**This is part 4 of the list of changes for pipeline comments**

- Adds pipeline level comments
  - Adds a pipeline comments button in the pipeline actions 
  - User can show/hide the pipeline level comments from the button
  - The popover allows the user to browse/edit/add new comments at the pipeline level
- Fixes comments to show in deployed pipeline view.

### CODE REVIEW

File: cdap-ui/app/cdap/components/PipelineCanvasActions/PipelineCommentsActionBtn.tsx [hunk 1]
```diff
@@ -71,34 +74,44 @@ const useStyle = makeStyles<Theme, Partial<IPipelineCommentsActionBtnProps>>((th
...
+    setLocalToggle(false);
+    setAnchorEl(null);
+    onCommentsToggle(true);
   };
+
   React.useEffect(() => {
-    if (typeof doesStagesHaveComments === 'boolean') {
-      setShowMarker(doesStagesHaveComments);
+    if (Array.isArray(comments) && comments.length > 0) {
+      setShowMarker(comments.length > 0);
```
Comment (njbriggs) [inline cdap-ui/app/cdap/components/PipelineCanvasActions/PipelineCommentsActionBtn.tsx]: Just make this `setShowMarker(true)`? Or even make this whole block `setShowMarker(Array.isArray(comments) && comments.length > 0)`?

**Review by njbriggs — APPROVED** [review]
LGTM - one little thing to look at.

File: cdap-ui/app/cdap/components/PipelineCanvasActions/PipelineComments.tsx [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    if (typeof onChange === 'function') {
+      localComments[id] = updatedComment;
+      setLocalComments(filterEmptyComments(localComments));
+      onChange(localComments);
+    }
+  };
+  const onDelete = (id) => {
+    if (typeof onChange === 'function') {
+      setLocalComments([...comments.slice(0, id), ...comments.slice(id + 1)]);
+      onChange([...comments.slice(0, id), ...comments.slice(id + 1)]);
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/PipelineCanvasActions/PipelineComments.tsx]: extract the array so we don't need to recreate it?

File: cdap-ui/app/cdap/components/PipelineCanvasActions/PipelineComments.tsx [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+                enabled: true,
+                element: arrowRef,
+              },
+              offset: {
+                enabled: true,
+                offset: '200%, 10%',
+              },
+            }}
+          >
+            <div className={classes.arrow} ref={setArrowRef}></div>
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/PipelineCanvasActions/PipelineComments.tsx]: nit: remove the `</div>`

File: cdap-ui/app/directives/dag-plus/my-dag.html [hunk 4]
```diff
@@ -100,12 +100,13 @@
...
-    on-comments-toggle="DAGPlusPlusCtrl.toggleComments"
-    toggle="DAGPlusPlusCtrl.commentsState"
+    tooltip="'Show/Hide Pipeline Comments'"
+    on-comments-toggle="DAGPlusPlusCtrl.togglePipelineComments"
+    comments="DAGPlusPlusCtrl.pipelineComments"
+    on-change="DAGPlusPlusCtrl.setPipelineComments"
+    disabled='DAGPlusPlusCtrl.isDisabled'
   ></pipeline-comments-action-btn>
-</div>
+  </div>
```
Comment (elfenheart) [inline cdap-ui/app/directives/dag-plus/my-dag.html]: nit: the indentation seemed off?

**Review by elfenheart — APPROVED** [review]

**Comment by ajainarayanan** [comment]
Will cherry-pick the commit to the base feature branch.

**Pull request closed without merging**